cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java

```diff
@@ -234,54 +242,107 @@ public void dropAllForFlow(FlowId flowId) throws Exception {
 
   @Override
   public void upgrade() throws Exception {
-    // For each queue config table and queue data table in each namespace, perform an upgrade
-    for (final NamespaceMeta namespaceMeta : namespaceQueryAdmin.list()) {
-      impersonator.doAs(namespaceMeta.getNamespaceId(), new Callable<Void>() {
-        @Override
-        public Void call() throws Exception {
-          upgradeQueues(namespaceMeta);
-          return null;
+    int numThreads = 10; // TODO: Make this configurable
+    final ExecutorService executor =
+      Executors.newFixedThreadPool(numThreads,
+                                   new ThreadFactoryBuilder().setNameFormat("hbase-cmd-executor-%d").build());
+    try {
+      final Map<TableId, Future<?>> allFutures = new HashMap<>();
+      // For each queue config table and queue data table in each namespace, perform an upgrade
+      for (final NamespaceMeta namespaceMeta : namespaceQueryAdmin.list()) {
+        impersonator.doAs(namespaceMeta.getNamespaceId(), new Callable<Void>() {
+          @Override
+          public Void call() throws Exception {
+            Map<TableId, Future<?>> futures = upgradeQueues(namespaceMeta, executor);
+            allFutures.putAll(futures);
+            return null;
+          }
+        });
+      }
+
+      // Wait for the queue upgrades to complete
+      Map<TableId, Throwable> failed = waitForUpgrade(allFutures);
+      if (!failed.isEmpty()) {
+        for (Map.Entry<TableId, Throwable> entry : failed.entrySet()) {
+          LOG.error("Failed to upgrade queue table {}", entry.getKey(), entry.getValue());
         }
-      });
+        throw new Exception(String.format("Error upgrading queue tables. %s of %s failed",
+                                          failed.size(), allFutures.size()));
+      }
+    } finally {
+      try {
+        executor.shutdown();
+        executor.awaitTermination(5, TimeUnit.SECONDS);
```

Comment:
When reaching here, there are only two cases:

1. All futures completed, hence there should be nothing running in the executor, hence calling `executor.shutdownNow()` should be good enough.
2. This thread get interrupted such that the `waitForUpgrade` throws `InterruptedException` on the `Future.get` call. If that's the case, the `awaitTermination` call will alway throw `InterruptedException`.

I think we should either not let user to interrupt until the upgrade is done (by using a non-interruptable future get via `Uninterruptibles.getUninterruptibly`), or just call `executor.shutdownNow` with the threads as daemon thread so that the user can kill the upgrade at anything without blocking.